### PR TITLE
Query: Fix #4588 - Query :: SelectMany-GroupJoin-DefaultIfEmpty isn't…

### LINKED ISSUE #4588: Query :: SelectMany-GroupJoin-DefaultIfEmpty isn't being lifted into SQL for some complex queries, resulting in extensive client-side evaluation
This is mostly visible in scenarios with optional navigations, e.g.

```
from c in ctx.Customers
where c.Detail.Name == "Foo"
select c;
```

this gets translated into:

```
from c in ctx.Customers
join d in ctx.Details on c.Id equals d.CustomerId into grouping
from d in grouping.DefaultIfEmpty()
where d.Name == "Foo"
select c;
```

however the filter will be performed on the client resulting in unnecessary data being pulled from the database

### PR DESCRIPTION
**Pull request by anpete**
… being lifted into SQL for some complex queries, resulting in extensive client-side evaluation.

Adds relational GroupJoin/DefaultIfEmpty elimination.

### CODE REVIEW

**Comment by anpete** [comment]
@maumar @smitpatel

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -1007,6 +1021,28 @@ protected override Expression VisitQuerySourceReference(QuerySourceReferenceExpr
...
+                {
+                    var entityType
+                        = _queryModelVisitor.QueryCompilationContext.Model
+                            .FindEntityType(joinClause.ItemType);
+
+                    if (entityType != null)
+                    {
+                        return Visit(
+                            EntityQueryModelVisitor.CreatePropertyExpression(
+                                expression, entityType.FindPrimaryKey().Properties[0]));
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: Why PK & only first property

**Comment by smitpatel** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: This is the case when a bare QSRE is encountered E.g. Where(c => c != null && c.CustomerID == "ALFKI")

We only need to test a single PK column because none can be null.
